File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/execution/CqlEngineTest.java
package org.opencds.cqf.cql.engine.execution;

import static org.hamcrest.CoreMatchers.containsString;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import org.junit.jupiter.api.Test;
import org.opencds.cqf.cql.engine.debug.DebugMap;
import org.opencds.cqf.cql.engine.exception.CqlException;

class CqlEngineTest extends CqlTestBase {

    @Test
    void debugMap() {

        // The specific library isn't important, just that it has a debug map
        var debugMap = new DebugMap();
        debugMap.setIsLoggingEnabled(true);
        var results = engine.evaluate(toElmIdentifier("CqlEngineTest"), null, null, null, debugMap);

        var libraryDebug = results.getDebugResult()
                .getLibraryResults()
                .get("CqlEngineTest")
                .getResults();

        assertNotNull(libraryDebug);

        // Find the debug result for the AnInteger expression
        // It's indexed by location
        var result = libraryDebug.keySet().stream()
                .filter(e -> e.getLocator().equals("6:1-6:21"))
                .findFirst();

        assertTrue(result.isPresent());

        var debugResult = libraryDebug.get(result.get());
        assertEquals(1, debugResult.size());
    }

    @Test
    public void invalidCql() {
        var exception = assertThrows(CqlException.class, () -> engine.evaluate(toElmIdentifier("Invalid")));
        assertThat(exception.getMessage(), containsString("Library Invalid loaded, but had errors:"));
    }

    @Test
    public void hedisCompatibility() {
        var libraryResult = engine.evaluate(toElmIdentifier("HedisCompatibilityTest"));
        var result = libraryResult.expressionResults.get("QuantityListIncludes").value();
        assertFalse((Boolean) result);

        result = libraryResult.expressionResults.get("ReturnUnspecified").value();
        assertInstanceOf(List.class, result);
        assertEquals(2, ((List<?>) result).size());

        result = libraryResult.expressionResults.get("ReturnAll").value();
        assertInstanceOf(List.class, result);
        assertEquals(5, ((List<?>) result).size());

        result = libraryResult.expressionResults.get("ReturnDistinct").value();
        assertInstanceOf(List.class, result);
        assertEquals(2, ((List<?>) result).size());

        result = libraryResult.expressionResults.get("Test Null Tuple").value();
        assertNull(result);

        engine.getState().getEngineOptions().add(CqlEngine.Options.EnableHedisCompatibilityMode);
        libraryResult = engine.evaluate(toElmIdentifier("HedisCompatibilityTest"));
        // equivalent semantics for lists
        result = libraryResult.expressionResults.get("QuantityListIncludes").value();
        assertTrue((Boolean) result);

        // no "distinct" behavior for lists
        result = libraryResult.expressionResults.get("ReturnUnspecified").value();
        assertInstanceOf(List.class, result);
        assertEquals(5, ((List<?>) result).size());

        result = libraryResult.expressionResults.get("ReturnAll").value();
        assertInstanceOf(List.class, result);
        assertEquals(5, ((List<?>) result).size());

        result = libraryResult.expressionResults.get("ReturnDistinct").value();
        assertInstanceOf(List.class, result);
        assertEquals(5, ((List<?>) result).size());

        result = libraryResult.expressionResults.get("Test Null Tuple").value();
        assertNull(result);
    }
}

File: Src/java/engine/src/test/kotlin/org/opencds/cqf/cql/engine/execution/CqlEngineTest.kt
package org.opencds.cqf.cql.engine.execution

import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.debug.DebugMap
import org.opencds.cqf.cql.engine.debug.DebugResultEntry
import org.opencds.cqf.cql.engine.exception.CqlException

internal class CqlEngineTest : CqlTestBase() {
    @Test
    fun debugMap() {
        // The specific library isn't important, just that it has a debug map

        val debugMap = DebugMap()
        debugMap.isLoggingEnabled = true
        val results = engine.evaluate(toElmIdentifier("CqlEngineTest"), null, null, null, debugMap)

        val libraryDebug = results.debugResult!!.libraryResults["CqlEngineTest"]!!.results

        Assertions.assertNotNull(libraryDebug)

        // Find the debug result for the AnInteger expression
        // It's indexed by location
        val result =
            libraryDebug.keys.stream().filter { e -> e!!.locator == "6:1-6:21" }.findFirst()

        Assertions.assertTrue(result.isPresent)

        val debugResult: MutableList<DebugResultEntry?> = libraryDebug[result.get()]!!
        Assertions.assertEquals(1, debugResult.size)
    }

    @Test
    fun invalidCql() {
        val versionedIdentifier = toElmIdentifier("Invalid")
        val exception =
            Assertions.assertThrows(CqlException::class.java) {
                engine.evaluate(versionedIdentifier)
            }
        MatcherAssert.assertThat<String?>(
            exception.message,
            CoreMatchers.containsString("Library Invalid loaded, but had errors:"),
        )
    }

    @Test
    fun withVersion() {
        val versionedIdentifierWithVersion =
            toElmIdentifier("LibraryWithVersion").withVersion("1.0.0")
        val versionedIdentifierNoVersion = toElmIdentifier("LibraryWithVersion")

        val singleLibResult = engine.evaluate(versionedIdentifierWithVersion)
        Assertions.assertNotNull(singleLibResult)

        val multiLibResults = engine.evaluate(listOf(versionedIdentifierWithVersion))
        Assertions.assertNotNull(multiLibResults)
        Assertions.assertNotNull(multiLibResults.getResultFor(versionedIdentifierNoVersion))
    }

    @Test
    fun hedisCompatibility() {
        var libraryResult = engine.evaluate(toElmIdentifier("HedisCompatibilityTest"))
        var result = libraryResult.expressionResults["QuantityListIncludes"]!!.value()
        Assertions.assertFalse((result as Boolean?)!!)

        result = libraryResult.expressionResults["ReturnUnspecified"]!!.value()
        Assertions.assertInstanceOf(MutableList::class.java, result)
        Assertions.assertEquals(2, (result as MutableList<*>).size)

        result = libraryResult.expressionResults["ReturnAll"]!!.value()
        Assertions.assertInstanceOf(MutableList::class.java, result)
        Assertions.assertEquals(5, (result as MutableList<*>).size)

        result = libraryResult.expressionResults["ReturnDistinct"]!!.value()
        Assertions.assertInstanceOf(MutableList::class.java, result)
        Assertions.assertEquals(2, (result as MutableList<*>).size)

        result = libraryResult.expressionResults["Test Null Tuple"]!!.value()
        Assertions.assertNull(result)

        engine.state.engineOptions.add(CqlEngine.Options.EnableHedisCompatibilityMode)
        libraryResult = engine.evaluate(toElmIdentifier("HedisCompatibilityTest"))
        // equivalent semantics for lists
        result = libraryResult.expressionResults["QuantityListIncludes"]!!.value()
        Assertions.assertTrue((result as Boolean?)!!)

        // no "distinct" behavior for lists
        result = libraryResult.expressionResults["ReturnUnspecified"]!!.value()
        Assertions.assertInstanceOf(MutableList::class.java, result)
        Assertions.assertEquals(5, (result as MutableList<*>).size)

        result = libraryResult.expressionResults["ReturnAll"]!!.value()
        Assertions.assertInstanceOf(MutableList::class.java, result)
        Assertions.assertEquals(5, (result as MutableList<*>).size)

        result = libraryResult.expressionResults["ReturnDistinct"]!!.value()
        Assertions.assertInstanceOf(MutableList::class.java, result)
        Assertions.assertEquals(5, (result as MutableList<*>).size)

        result = libraryResult.expressionResults["Test Null Tuple"]!!.value()
        Assertions.assertNull(result)
    }
}
